### PR TITLE
feat: Balancer V2 load and cache top pools by num swaps on startup [TKR-96]

### PR DESCRIPTION
**Pull request by kimpers**
## Description
This implements a slightly hacky solution to pre-populate the Balancer V2 pool cache on start-up so we avoid having empty a cache the first time we try to sample a token pair. Once Balance releases a new version of SOR that is compatible with both V1 and V2 we can remove the copy-pasta and clean this up 😄 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts [hunk 1]
```diff
@@ -29,8 +70,74 @@ export class BalancerV2PoolsCache extends PoolsCache {
...
+                        this._warningLogger(err, `Failed to load Balancer V2 top pools`);
+                        // soldier on
+                    }
+                }
+            }
+        });
+    }
     protected async _fetchPoolsForPairAsync(takerToken: string, makerToken: string): Promise<Pool[]> {
-        const query = `
+        const query = gql`
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:140]: oh, never seen this syntax before, how does this work?

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts [hunk 2]
```diff
@@ -1,19 +1,60 @@
...
         private readonly subgraphUrl: string = BALANCER_V2_SUBGRAPH_URL,
         private readonly maxPoolsFetched: number = BALANCER_MAX_POOLS_FETCHED,
+        private readonly _topPoolsFetched: number = BALANCER_TOP_POOLS_FETCHED,
+        private readonly _warningLogger: LogFunction = DEFAULT_WARNING_LOGGER,
         cache: { [key: string]: CacheValue } = {},
     ) {
         super(cache);
+        void this._loadTopPoolsAsync();
+        // Reload the top pools every 12 hours
+        setInterval(async () => void this._loadTopPoolsAsync(), ONE_DAY_MS / 2);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:57]: why?

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts [hunk 3]
```diff
@@ -29,8 +70,74 @@ export class BalancerV2PoolsCache extends PoolsCache {
...
+                    }
+                    try {
+                        // The list of pools must be relevant to `from` and `to`  for `parsePoolData`
+                        const [poolData] = parsePoolData({ [pool.id]: pool as any }, from, to);
+                        fromToPools[from][to].push(
+                            BalancerV2PoolsCache._parseSubgraphPoolData(poolData[pool.id], from, to),
+                        );
+                        // Cache this as we progress through
+                        const expiresAt = Date.now() + this._cacheTimeMs;
+                        this._cachePoolsForPair(from, to, fromToPools[from][to], expiresAt);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:130]: `this` is not clobbered by the `forEach()`?

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts [hunk 4]
```diff
@@ -29,8 +70,74 @@ export class BalancerV2PoolsCache extends PoolsCache {
...
+        pools.forEach(pool => {
+            const { tokensList } = pool;
+            for (const from of tokensList) {
+                for (const to of tokensList.filter(t => t.toLowerCase() !== from.toLowerCase())) {
+                    if (!fromToPools[from]) {
+                        fromToPools[from] = {};
+                    }
+                    if (!fromToPools[from][to]) {
+                        fromToPools[from][to] = [];
+                    }
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:121]: perhaps
```suggestion
                    fromToPools[from] = fromToPools[from] || {};
                    fromToPools[from][to] = fromToPools[from][to] || {};
```

**Review by dorothy-zbornak — APPROVED** [review]
:+1: :eye: :lips: :eye: :+1:

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts (hunk 2 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:57]: If it wasn't discovered by the last deployment (on init), then the only way a pool can be discovered is through natural search. 

We want a method to discover decently liquid pools that users may not naturally search. Maybe a liquid BAL/USDT pool appears but no one ever searches BAL/USDT, they always search BAL/USDC.  We don't go searching on Multihop intermediary hops automatically.

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts (hunk 1 above)
Comment (kimpers) [reply packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:140]: I think it's just a tagged string template so it doesn't do so much. However, now that the editor understands that it's a GQL query you can get prettier formatting and some basic linting :)

**Pull request merged**